quickwit/quickwit-lambda/src/searcher/warp_lambda.rs

```diff
@@ -0,0 +1,173 @@
+    Response, Service,
+};
+use mime_guess::{mime, Mime};
+use once_cell::sync::Lazy;
+use tracing::{info_span, Instrument};
+use warp::hyper::Body as WarpBody;
+pub use {lambda_http, warp};
+
+pub type WarpRequest = warp::http::Request<warp::hyper::Body>;
+pub type WarpResponse = warp::http::Response<warp::hyper::Body>;
+
+pub async fn run<'a, S>(service: S) -> Result<(), LambdaError>
+where
+    S: Service<WarpRequest, Response = WarpResponse, Error = Infallible> + Send + 'a,
+    S::Future: Send + 'a,
+{
+    lambda_runtime::run(Adapter::from(WarpAdapter::new(service))).await
+}
+
+#[derive(Clone)]
+pub struct WarpAdapter<'a, S>
+where
+    S: Service<WarpRequest, Response = WarpResponse, Error = Infallible>,
+    S::Future: Send + 'a,
+{
+    warp_service: S,
+    _phantom_data: PhantomData<&'a WarpResponse>,
+}
+
+impl<'a, S> WarpAdapter<'a, S>
+where
+    S: Service<WarpRequest, Response = WarpResponse, Error = Infallible>,
+    S::Future: Send + 'a,
+{
+    pub fn new(warp_service: S) -> Self {
+        Self {
+            warp_service,
+            _phantom_data: PhantomData,
+        }
+    }
+}
+
+static PLAINTEXT_MIMES: Lazy<HashSet<Mime>> = Lazy::new(|| {
+    vec![
+        mime::APPLICATION_JAVASCRIPT,
+        mime::APPLICATION_JAVASCRIPT_UTF_8,
+        mime::APPLICATION_JSON,
+    ]
+    .into_iter()
+    .collect()
+});
+
+async fn warp_body_as_lambda_body(
+    warp_body: WarpBody,
+    parts: &Parts,
+) -> Result<LambdaBody, LambdaError> {
+    // Concatenate all bytes into a single buffer
+    let raw_bytes = warp::hyper::body::to_bytes(warp_body).await?;
+
+    // Attempt to determine the Content-Type
+    let content_type: Option<&HeaderValue> = parts.headers.get("Content-Type");
+    let content_encoding: Option<&HeaderValue> = parts.headers.get("Content-Encoding");
+
+    // If Content-Encoding is present, assume compression
+    // If Content-Type is not present, don't assume is a string
+    let body = if let (Some(typ), None) = (content_type, content_encoding) {
+        let typ = typ.to_str()?;
+        let m = typ.parse::<Mime>()?;
+        if PLAINTEXT_MIMES.contains(&m) || m.type_() == mime::TEXT {
+            Some(String::from_utf8(raw_bytes.to_vec()).map(LambdaBody::Text)?)
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+
+    // Not a text response, make binary
+    Ok(body.unwrap_or_else(|| LambdaBody::Binary(raw_bytes.to_vec())))
+}
+
+impl<'a, S> Service<Request> for WarpAdapter<'a, S>
+where
+    S: Service<WarpRequest, Response = WarpResponse, Error = Infallible> + 'a,
+    S::Future: Send + 'a,
+{
+    type Response = Response<LambdaBody>;
+    type Error = LambdaError;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'a>>;
+
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        core::task::Poll::Ready(Ok(()))
```

Comment:
why not call warp_service's own poll_ready?
```suggestion
    fn poll_ready(&mut self, ctx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.warp_service.poll_ready(ctx)
            .map_err(|err| match err {})
```

Reply:
This code comes from the forked https://github.com/aslamplr/warp_lambda. I think the rational is that when running on Lambda, we don't want the backpressure mechanism offered by `poll_ready` on the service trait. We have only one caller at a time. But I agree that we should make sure that the warp_service is ready, otherwise as per the `call` [method's spec](https://docs.rs/tower/latest/tower/trait.Service.html#tymethod.call) it's allowed to panic